### PR TITLE
Security2

### PR DESCRIPTION
**Pull request by VasylAzarov**
Implemented scripts for bd:
creating roles , user_roles tables;
inserting books, users, roles, user roles;
assign roles to users without it |
add and set spring security | add user roles | add checking role to endpoints

### CODE REVIEW

File: src/main/java/dev/vasyl/proj/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@RequiredArgsConstructor
+@EnableMethodSecurity
+@Configuration
+public class SecurityConfig {
+
+    private final CustomUserDetailsService userDetailsService;
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/config/SecurityConfig.java]:
```suggestion
    private final UserDetailsService userDetailsService;
```

File: src/main/java/dev/vasyl/proj/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers(
+                                        "/auth/**",
+                                        "/error",
+                                        "/swagger-ui/**",
+                                        "/v3/api-docs/**",
+                                        "/swagger-resources/**",
+                                        "/configuration/**",
+                                        "/webjars/**")
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/config/SecurityConfig.java]:
```suggestion
                                        "/auth/**",
                                        "/error",
                                        "/swagger-ui/**",
                                        "/v3/api-docs/**")
```

File: src/main/java/dev/vasyl/proj/controller/BookController.java [hunk 3]
```diff
@@ -29,45 +30,50 @@
...
+            description = "Get Book page by page params or/and sort params."
+                    + " Available for User and Admin roles")
     public Page<BookDto> getAll(@ParameterObject
                                     @PageableDefault(size = 20, sort = "title",
                                             direction = Sort.Direction.ASC)
                                             Pageable pageable) {
         return bookService.findAll(pageable);
     }
 
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/controller/BookController.java]: in your implementation, the user with the role ADMIN has a USER role by default so you can validate only the USER role here

File: src/main/java/dev/vasyl/proj/model/User.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Column(nullable = false)
+    private String firstName;
+
+    @Column(nullable = false)
+    private String lastName;
+
+    private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/model/User.java]: Don’t  use EAGER fetch type

File: src/main/java/dev/vasyl/proj/model/User.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "users")
+public class User implements UserDetails {
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/model/User.java:28]: make user entity soft-deleted.

File: src/main/java/dev/vasyl/proj/model/User.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/dev/vasyl/proj/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -36,6 +42,11 @@ public BookDto findById(long id) {
         return bookMapper.toDto(book);
     }
 
+    @Override
+    public boolean existByIsbn(String isbn) {
+        return bookRepository.existsByIsbn(isbn);
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/BookServiceImpl.java]: do you really need this method? you can call the repository method directly

File: src/main/java/dev/vasyl/proj/service/impl/RoleServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        return roleRepository.save(role);
+    }
+
+    @Override
+    public Role findByName(RoleName roleName) {
+        return roleRepository.findByName(roleName)
+                .orElseThrow(() -> new EntityNotFoundException("Role not found: " + roleName));
+    }
+}
+
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/RoleServiceImpl.java]:
```suggestion
```

File: src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.existsByEmail(requestDto.getEmail())) {
+            throw new RegistrationException("user with email ["
+                    + requestDto.getEmail()
+                    + "] already exist");
+        }
+        User user = userMapper.toModel(requestDto);
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java:35]: Let’s  add the default USER role

File: src/main/java/dev/vasyl/proj/sucurity/CustomUserDetailsService.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new EntityNotFoundException("Cant find user by email"));
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/sucurity/CustomUserDetailsService.java]:
```suggestion
```

File: src/main/java/dev/vasyl/proj/sucurity/CustomUserDetailsService.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new EntityNotFoundException("Cant find user by email"));
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/sucurity/CustomUserDetailsService.java]: add email parameter

File: src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            Field confirmPasswordField
+                    = value.getClass().getDeclaredField(this.confirmPasswordField);
+            passwordField.setAccessible(true);
+            confirmPasswordField.setAccessible(true);
+            String password = (String) passwordField.get(value);
+            String confirmPassword = (String) confirmPasswordField.get(value);
+
+            if (Objects.equals(password, confirmPassword)) {
+                return true;
+            }
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java]:
```suggestion
```

File: src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+            return password.equals(confirmPassword);
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java]:
```suggestion
            return Objects.equals(password, confirmPassword);
```

File: src/main/resources/db/changelog/changes/04-create-user_roles-table.yaml [hunk 14]
```diff
@@ -0,0 +1,31 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-user-roles-table
+      author: vasyl
+      runOnChange: true
+      changes:
+        - createTable:
+            tableName: user_roles
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/04-create-user_roles-table.yaml]:
```suggestion
            tableName: users_roles
```

File: src/main/resources/db/changelog/changes/05-insert-books-if-not-exist.yaml [hunk 15]
```diff
@@ -0,0 +1,76 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-books-if-not-exist
+      author: vasyl
+      runOnChange: true
+      preConditions:
+        - onFail: MARK_RAN
+          sqlCheck:
+            sql: "SELECT COUNT(*) FROM books"
+            expectedResult: 0
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/05-insert-books-if-not-exist.yaml]: do you really need this check? Remove or refactor it;)

File: src/main/resources/db/changelog/changes/07-insert-users-if-not-exist.yaml [hunk 16]
```diff
@@ -0,0 +1,67 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-users-if-not-exist
+      author: vasyl
+      runOnChange: true
+      preConditions:
+        - onFail: MARK_RAN
+          sqlCheck:
+            sql: "SELECT COUNT(*) FROM users"
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/07-insert-users-if-not-exist.yaml]: the same. select by email or remove this condition

File: src/main/resources/db/changelog/changes/08-insert-admin-user-if-none-exist.yaml [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+      runOnChange: true
+      preConditions:
+        - onFail: MARK_RAN
+        - sqlCheck:
+            expectedResult: 0
+            sql: |
+              SELECT COUNT(*)
+              FROM user_roles ur
+              JOIN roles r ON ur.role_id = r.id
+              WHERE r.name = 'ADMIN'
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/08-insert-admin-user-if-none-exist.yaml]: the same issue

File: src/main/resources/db/changelog/changes/09-assign-user-role-to-users-without-roles.yaml [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+              INSERT INTO user_roles (user_id, role_id)
+              SELECT u.id, r.id
+              FROM users u
+              CROSS JOIN roles r
+              WHERE r.name = 'USER'
+              AND NOT EXISTS (
+                SELECT 1
+                FROM user_roles ur
+                WHERE ur.user_id = u.id
+              );
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/09-assign-user-role-to-users-without-roles.yaml]: Let’s remove it.

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution!

File: src/main/java/dev/vasyl/proj/model/Role.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Table(name = "roles")
+@Getter
+@Setter
+public class Role {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/model/Role.java]:
```suggestion
public class Role implements GrantedAuthority {
```

File: src/main/java/dev/vasyl/proj/model/User.java [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+@Getter
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/model/User.java:26]: check if `@Where` is deprecated or not

File: src/main/java/dev/vasyl/proj/model/User.java [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+    private String shippingAddress;
+
+    @ManyToMany
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/dev/vasyl/proj/model/User.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of();
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/model/User.java]:
```suggestion
        return roles;
```

File: src/main/java/dev/vasyl/proj/repository/UserRepository.java [hunk 23]
```diff
@@ -0,0 +1,16 @@
+package dev.vasyl.proj.repository;
+
+import dev.vasyl.proj.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/repository/UserRepository.java]: Be consistent, if you use `@Repository` use it everywhere or remove

File: src/main/java/dev/vasyl/proj/service/RoleService.java [hunk 24]
```diff
@@ -0,0 +1,10 @@
+package dev.vasyl.proj.service;
+
+import dev.vasyl.proj.model.Role;
+import dev.vasyl.proj.model.RoleName;
+
+public interface RoleService {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/RoleService.java]: remove this service

File: src/main/java/dev/vasyl/proj/service/impl/RoleServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+import dev.vasyl.proj.model.Role;
+import dev.vasyl.proj.model.RoleName;
+import dev.vasyl.proj.repository.RoleRepository;
+import dev.vasyl.proj.service.RoleService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class RoleServiceImpl implements RoleService {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/impl/RoleServiceImpl.java]: remove

File: src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Role role = roleRepository.findByName(RoleName.USER).orElseThrow(
+                () -> new EntityNotFoundException("Error when set user role"));
+        user.setRoles(Set.of(role));
+        user = userRepository.save(user);
+        return userMapper.toUserResponse(user);
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
        return userMapper.toUserResponse(user);
```

File: src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class UserDetailsService
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java]:
```suggestion
public class CustomUserDetailsService implements UserDetailsService {
```

File: src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java [hunk 28]
```diff
@@ -0,0 +1,41 @@
+package dev.vasyl.proj.sucurity;
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java]:
```suggestion
package dev.vasyl.proj.security;
```

File: src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+                        ));
+                    }
+                    return new org.springframework.security.core.userdetails.User(
+                            user.getEmail(),
+                            user.getPassword(),
+                            authorities
+                    );
+                })
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Cant find user by email: [" + email + "]"));
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java]:
```suggestion
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new EntityNotFoundException(.......
```

File: src/main/resources/db/changelog/changes/05-insert-books.yaml [hunk 30]
```diff
@@ -0,0 +1,96 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-books-1
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/05-insert-books.yaml]: don't need insert books to DB

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's resolve conflicts in branch

see comments

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**